core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1607,11 +1616,18 @@ private[spark] class DAGScheduler(
                   case mapStage: ShuffleMapStage =>
                     val numMissingPartitions = mapStage.findMissingPartitions().length
                     if (numMissingPartitions < mapStage.numTasks) {
-                      // TODO: support to rollback shuffle files.
-                      // Currently the shuffle writing is "first write wins", so we can't re-run a
-                      // shuffle map stage and overwrite existing shuffle files. We have to finish
-                      // SPARK-8029 first.
-                      abortStage(mapStage, generateErrorMessage(mapStage), None)
+                      if (sc.getConf.get(config.SHUFFLE_USE_OLD_FETCH_PROTOCOL)) {
+                        val reason = "A shuffle map stage with indeterminate output was failed " +
+                          "and retried. However, Spark can only do this while using the new " +
+                          "shuffle block fetching protocol. Please check the config " +
+                          "'spark.shuffle.useOldFetchProtocol', see more detail in " +
+                          "SPARK-27665 and SPARK-25341."
+                        abortStage(mapStage, reason, None)
+                      } else {
+                        logInfo(s"The indeterminate stage $mapStage will be resubmitted," +
+                          " the stage self and all indeterminate parent stage will be" +
+                          " rollback and whole stage rerun.")
```

Comment:
I find this msg a little confusing.  You're not rolling back the parents of the indeterminate stages, you're rolling back its children, right?  Also, you'll be logging this for every stage you are rolling back, which should include the entire set of stages which actually get rolled back, right?  I think it might be simpler to have it just log once with the set of stages that get rolled back.

Reply:
Yeah, the answer for both questions is yes. Just log once done in aa3a409.